packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1684,8 +1687,10 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     } else {
       newRoute.didChangeNext(null);
     }
-    if (index > 0)
+    if (index > 0) {
       _history[index - 1].didChangeNext(newRoute);
+      newRoute.didChangePrevious(_history[index - 1]);
+    }
     for (NavigatorObserver observer in widget.observers)
       observer.didReplace(newRoute: newRoute, oldRoute: oldRoute);
     oldRoute.dispose();
```

Comment:
since this is technically a breaking change (albeit one that should only make things better), please go through the breaking change process as documented on the design principles page.